cpp/src/fil/internal.cuh

```diff
@@ -322,42 +324,44 @@ struct categorical_sets {
   // arrays from each node ID are concatenated first, then from all categories
   const uint8_t* bits = nullptr;
   // largest matching category in the model, per feature ID
-  const int* max_matching       = nullptr;
-  std::size_t bits_size         = 0;
-  std::size_t max_matching_size = 0;
+  const float* fid_cats = nullptr;
+  std::size_t bits_size = 0;
+  // either 0 or num_cols. When 0, indicates intended empty array.
+  std::size_t fid_cats_size = 0;
 
   __host__ __device__ __forceinline__ bool cats_present() const
   {
-    // If this is constructed from cat_sets_owner, will return true
-    // default-initialized will return false
-    // Defining edge case: there are categorical nodes, but all have max_matching == -1
-    // (all categorical nodes are empty). node.thresh() would have returned 0.0f
-    // and the branch condition wouldn't have always been false (i.e branched left).
-    // Alternatively, we could have converted all empty categorical nodes to
-    // NAN-threshold numerical nodes.
-    return max_matching != nullptr;
+    // If this is constructed from cat_sets_owner, will return true; but false by default
+    // We have converted all empty categorical nodes to NAN-threshold numerical nodes.
+    return fid_cats != nullptr;
   }
 
   // set count is due to tree_idx + node_within_tree_idx are both ints, hence uint32_t result
   template <typename node_t>
-  __host__ __device__ __forceinline__ int category_matches(node_t node, int category) const
+  __host__ __device__ __forceinline__ int category_matches(node_t node, float category) const
   {
     // standard boolean packing. This layout has better ILP
     // node.set() is global across feature IDs and is an offset (as opposed
     // to set number). If we run out of uint32_t and we have hundreds of
     // features with similar categorical feature count, we may consider
     // storing node ID within nodes with same feature ID and look up
-    // {.max_matching, .first_node_offset} = ...[feature_id]
-    return category <= max_matching[node.fid()] && fetch_bit(bits + node.set(), category);
+    // {.fid_cats, .first_node_offset} = ...[feature_id]
+
+    /* category < 0.0f or category > INT_MAX is equivalent to out-of-dictionary category
+    (not matching, branch left). -0.0f represents category 0.
+    If (float)(int)category != category, we will discard the fractional part.
+    E.g. 3.8f represents category 3 regardless of fid_cats value.
+    FIL will reject a model where an integer within [0, fid_cats] cannot be represented
+    precisely as a 32-bit float.
+    */
+    return category < fid_cats[node.fid()] && category >= 0.0f &&
+           fetch_bit(bits + node.set(), static_cast<int>(category));
```

Comment:
Am I interpreting it correctly that we're still doing zero-ward rounding here with that `static_cast`?

Reply:
yes